js.go

```diff
@@ -0,0 +1,1175 @@
+		return nil, err
+	}
+
+	var info JSApiConsumerResponse
+	if err := json.Unmarshal(resp.Data, &info); err != nil {
+		return nil, err
+	}
+	if info.Error != nil {
+		return nil, errors.New(info.Error.Description)
+	}
+	return info.ConsumerInfo, nil
+}
+
+func (m *Msg) checkReply() (*js, bool, error) {
+	if m.Reply == "" {
+		return nil, false, ErrMsgNoReply
+	}
+	if m == nil || m.Sub == nil {
+		return nil, false, ErrMsgNotBound
+	}
+	sub := m.Sub
+	sub.mu.Lock()
+	js := sub.jsi.js
+	isPullMode := sub.jsi.pull > 0
+	sub.mu.Unlock()
+
+	if js == nil {
+		return nil, false, ErrNotJSMessage
+	}
+	return js, isPullMode, nil
+}
+
+// ackReply handles all acks. Will do the right thing for pull and sync mode.
+func (m *Msg) ackReply(ackType []byte, sync bool) error {
+	js, isPullMode, err := m.checkReply()
+	if err != nil {
+		return err
+	}
+	if isPullMode {
+		if bytes.Equal(ackType, AckAck) {
+			err = js.nc.PublishRequest(js.apiSubj(m.Reply), m.Sub.Subject, AckNext)
+		} else if bytes.Equal(ackType, AckNak) || bytes.Equal(ackType, AckTerm) {
+			err = js.nc.PublishRequest(js.apiSubj(m.Reply), m.Sub.Subject, []byte("+NXT {\"batch\":1}"))
+		}
+		if sync && err == nil {
+			_, err = js.nc.Request(js.apiSubj(m.Reply), nil, defaultRequestWait)
+		}
+	} else if sync {
+		_, err = js.nc.Request(js.apiSubj(m.Reply), ackType, defaultRequestWait)
+	} else {
+		err = js.nc.Publish(js.apiSubj(m.Reply), ackType)
+	}
+	return err
+}
+
+// Acks for messages
+
+// Ack a message, this will do the right thing with pull based consumers.
+func (m *Msg) Ack() error {
+	return m.ackReply(AckAck, false)
+}
+
+// Ack a message and wait for a response from the server.
+func (m *Msg) AckSync() error {
+	return m.ackReply(AckAck, true)
+}
+
+// Nak this message, indicating we can not process.
+func (m *Msg) Nak() error {
+	return m.ackReply(AckNak, false)
+}
+
+// Term this message from ever being delivered regardless of MaxDeliverCount.
+func (m *Msg) Term() error {
+	return m.ackReply(AckTerm, false)
+}
+
+// Indicate that this message is being worked on and reset redelkivery timer in the server.
+func (m *Msg) InProgress() error {
+	return m.ackReply(AckProgress, false)
+}
+
+// Jetstream metadata associated with received messages.
+type MsgMetaData struct {
+	Consumer  uint64
+	Stream    uint64
+	Delivered uint64
+	Pending   uint64
+	Timestamp time.Time
+}
+
+func (m *Msg) MetaData() (*MsgMetaData, error) {
+	if _, _, err := m.checkReply(); err != nil {
+		return nil, err
+	}
+
+	const expectedTokens = 9
+	const btsep = '.'
+
+	tsa := [expectedTokens]string{}
+	start, tokens := 0, tsa[:0]
+	subject := m.Reply
+	for i := 0; i < len(subject); i++ {
+		if subject[i] == btsep {
+			tokens = append(tokens, subject[start:i])
+			start = i + 1
+		}
+	}
+	tokens = append(tokens, subject[start:])
+	if len(tokens) != expectedTokens || tokens[0] != "$JS" || tokens[1] != "ACK" {
+		return nil, ErrNotJSMessage
+	}
+
+	meta := &MsgMetaData{
+		Delivered: uint64(parseNum(tokens[4])),
```

Comment:
parseNum can return -1 which will panic when cast to uint64

Reply:
Actually just makes it a big number.

https://play.golang.org/p/Rxyg7lYmDv6

Reply:
Hmm swear I did the same play and it panicked lol. Don’t work in tbe same room as a 3 year old :P

Reply:
lol no worries, I knew it was ok b/c I hit that issue in JetStream's filestore ;)